src/util/std_types.h

```diff
@@ -1071,6 +1071,12 @@ class bitvector_typet : public typet
     set(ID_width, narrow_cast<long long>(width));
   }
 
+  mp_integer smallest() const;
+  mp_integer largest() const;
+  constant_exprt smallest_expr() const;
+  constant_exprt zero_expr() const;
+  constant_exprt largest_expr() const;
```

Comment:
This is brilliant refactoring, thank you! Yet can you please hide these in `pointer_typet`? A `pointer_typet` happens to be a `bitvector_typet`, but really they must not be used (not even `zero_expr`) on a `pointer_typet`.

Reply:
Ah, I overlooked that there are actually quite a few classes that inherit from `bitvector_typet`. Besides the three that were already mentioned, there are `bv_typet`, `fixedbv_typet`, `floatbv_typet`, `c_bit_field_typet` and `c_bool_typet`. However these functions will hit an `UNREACHABLE` if the type id is anything other than `ID_unsignedbv` or `ID_signedbv`. Would it make sense to introduce a new class that inherits from `bitvector_typet` and that both `signedbv_typet` and `unsignedbv_typet` inherit from? Any suggestions for what such a class should be called?

Reply:
smallest()/largest() only make sense for the integer ones -- so that could be integer_bitvector_typet. smallest_expr, zero_expr, largest_expr() also work on fixedbv and floatbv; these could be a numeric_bitvector_typet.

Reply:
That makes sense, thank you! I only created the `integer_bitvector_typet` class as all five methods were previously only defined for the signedbv/unsignedbv types, and I added some documentation and a comment mentioning that the methods returning an exprt would make sense for fixedbv/floatbv as well in case we need them there in the future.